Object oriented way to describe index settings and mapping

maijs opened the pull request:
This change adds two methods to `ElasticsearchIndexInterface` interface:
- `getIndexDefinition()` which is expected to be an instance of `IndexDefinition` class (not required in index plugins, but is strongly suggested)
- `getMappingDefinition()`which is expected to be an instance of `MappingDefinition` class (required to be defined in index plugins)

This PR breaks the backwards compatibility due to new additions to the interface.

By making index plugins explicitly define fields in an object-oriented way, integration with Elasticsearch Helper Views module can improved with automatic field and filter generation.

Elasticsearch index plugins are expected to return the following field definitions in `getMappingDefinition()` method. This is what has been defined in an array in `setup()` method right before sending to `putMapping()` client method before in legacy index plugins. Now `MappingsDefinition` contains a list of `FieldDefinition` which defines a field with additional parameters (options) if necessary.

```
  /**
   * {@inheritdoc}
   */
  public function getMappingDefinition() {
    $user_property = FieldDefinition::create('object')
      ->addProperty('id', FieldDefinition::create('integer'))
      ->addProperty('name', FieldDefinition::create('keyword'));

    return MappingsDefinition::create()
      ->addProperty('id', FieldDefinition::create('integer'))
      ->addProperty('uuid', FieldDefinition::create('keyword'))
      ->addProperty('title', FieldDefinition::create('text'))
      ->addProperty('status', FieldDefinition::create('keyword'))
      ->addProperty('user', $user_property);
  }
```

If Elasticsearch index plugin defines `getIndexDefinition()` and returns `IndexDefinition` instance, then `ElasticsearchIndexBase::setup` method will automatically create an index from given index definition.

```
/**
 * {@inheritdoc}
 */
public function getIndexDefinition() {
  // Get field mappings.
  $mappings = $this->getMappingDefinition();

  // Get index settings.
  $settings = SettingsDefinition::create()
    ->addOptions([
      'number_of_shards' => 1,
      'number_of_replicas' => 0,
    ]);

  $index_definition = IndexDefinition::create()
    ->setMappings($mappings)
    ->setSettings($settings);

  // If you are using Elasticsearch < 7, add the type to the index definition.
  $index_definition->setType($this->getTypeName([]));

  return $index_definition;
}
```

If plugin does not return `IndexDefinition` instance in `getIndexDefinition()` method or does not define the method, it's up to plugin's `setup` method to create an index. Still, it can use the mappings from its `getMappingDefinition()` method to setup the mapping:

```
/**
 * {@inheritdoc}
 */
public function setup() {
  $index_name = $this->getIndexName([]);

  if (!$this->client->indices()->exists(['index' => $index_name])) {
    $this->client->indices()->create([
      'index' => $index_name,
      'body' => [
        'settings' => [
           'number_of_shards' => 1,
           'number_of_replicas' => 0,
        ],
        'mappings' => $this->getMappingDefinition()->toArray(),
      ],
    ]);
  }
}
```

Example module is updated to provide code snippets how to use the new definition classes in Elasticsearch index plugins.

maijs committed:
Class ElasticsearchClientVersion added which returns full, major, minor and patch
version of the Elasticsearch client library.

maijs committed:
Phpdoc is updated in ElasticsearchIndexInterface to include the return type of the following
methods:
- index()
- get()
- delete()
- search()
- msearch()
- bulk()

maijs committed:
Elasticsearch related events added:
- Elasticsearch operation event which is used when common Elasticsearch operation
  is invoked - index, search, get, delete etc.
- Elasticsearch operation request event which is used before request to Elasticsearch
  is sent.

maijs committed:
Event subscribers are invoked when performing Elasticsearch operations.

When Elasticsearch index plugin receives an index-able object (entity or array) for indexing,
Elasticsearch operation event is emited to allow event subscribers to decide if object needs
to be indexed.

Before request is sent to Elasticserach, Elasticsearch operation request event is emited to
allow event subscribers to modify the request execution callback and/or its parameters.

maijs committed:
TypeEventSubscriber added which removes the "type" parameter from the request when interacting
with Elasticsearch version >= 7.

maijs committed:
Callable are renamed callback.

maijs committed:
Merge branch '8.x-7.x' into event-subscribers

* 8.x-7.x:
  Test scenarios for insert, update and delete are added.

maijs committed:
Elasticsearch index plugin instance is added to Elasticsearch operation and Elasticsearch
operation request events.

maijs committed:
Data type repository added.

- Data type class allows defining index mappings.

maijs committed:
Mapping definition, index setting definition and index definition classes added.

- Method getIndexDefinition() added to ElasticsearchIndexInterface which should
  return instance of IndexDefinition which describes the mapping and the settings
  of the index.

- Method getIndexMappings() added to ElasticsearchIndexInterface  which should
  return an instance of MappingsDefinition which describes the properties of the
  index.

maijs committed:
Merge branch '8.x-7.x' into index-definition

* 8.x-7.x:
  Updated elasticsearch_helper_example module to be Drupal 9 compatible
  Added a test for deferred indexing
  Elasticsearch PHP library availability is checked in runtime phase

maijs committed:
Elasticsearch indices are updated in "elasticsearch_helper_example" module.

- All index plugins are featuring getIndexMappings() method defined by the interface.

maijs committed:
Elasticsearch index plugin in test module updated to comply with the interface.

maijs committed:
TypeTrait added to provide compatibility with Elasticsearch < 7.

maijs committed:
Merge branch '8.x-7.x' into index-definition

* 8.x-7.x:
  Event emission is added to most common Elasticsearch operations

# Conflicts:
#	elasticsearch_helper.services.yml
#	src/Plugin/ElasticsearchIndexBase.php

misterjoonas commented on examples/elasticsearch_helper_example/src/Plugin/ElasticsearchIndex/SimpleNodeIndex.php:
```diff
@@ -16,7 +20,42 @@
+   */
+  public function getIndexMappings() {
+    $user_property = FieldDefinition::create('object')
+      ->addProperty('id', FieldDefinition::create('integer'))
```
Should this be uid instead, to make it more familiar with Drupal properties and avoid confusion with the document id? Technically can be whatever.

maijs replied:
Good point. Changed from `id` to `uid` for better clarity.

maijs committed:
User ID field is changed from "id" to "uid" in "simple_node_index" example plugin index.

maijs committed:
Class MappingsDefinition renamed to MappingDefinition.

maijs committed:
Interface method getIndexMappings() changed to getMappingDefinition().

mgalang commented on src/Elasticsearch/DefinitionBase.php:
```diff
@@ -0,0 +1,19 @@
+ * @see \Drupal\elasticsearch_helper\Elasticsearch\Index\MappingDefinition
+ * @see \Drupal\elasticsearch_helper\Elasticsearch\Index\SettingsDefinition
+ */
+class DefinitionBase {
```
Should this be an interface instead?

mgalang commented on src/Elasticsearch/ObjectTrait.php:
```diff
@@ -0,0 +1,99 @@
+   */
+  public function addOption($option, $value) {
+    $options = [$option => $value];
+    $this->validateOptions($options);
```
`$this->addOptions($options)` could just be called here instead.

maijs replied:
Good point!

mgalang commented on src/Plugin/ElasticsearchIndexBase.php:
```diff
@@ -106,11 +106,26 @@ protected function getEventDispatcher() {
+  /**
+   * {@inheritdoc}
+   */
+  public function getIndexDefinition() {
```
Maybe this empty method could be removed, then use a conditional check for the method from `setup()`?

`if (method_exists($this, 'getIndexDefinition') && $index_definition = $this->getIndexDefinition()) {`

maijs replied:
Method `getIndexDefinition()` is defined in `ElasticsearchIndexInterface` interface, therefore it is implemented in `ElasticsearchIndexBase`.

mgalang reviewed (approved):
I added a few suggestions for improvement, but otherwise is already great!

Awesome work here!

maijs committed:
Method "addOption()" in ObjectTrait is refactored.

mitrpaka commented on src/Elasticsearch/Index/FieldDefinition.php:
```diff
@@ -0,0 +1,271 @@
+   */
+  protected function validateOptions(array $options = []) {
+    if (isset($options['type'])) {
+      throw new \InvalidArgumentException(t('Data type should be added as an argument to the constructor of @class.', [
```
According to latest changes to coding standards, exception messages should not be translated. See https://www.drupal.org/docs/develop/coding-standards/php-exceptions#conventions

Thus, remove t() from all exceptions.

maijs committed:
Exception messages are not translated.

Exception messages should not be translated as per Drupal coding standards
(https://www.drupal.org/docs/develop/coding-standards/php-exceptions#conventions).

maijs committed:
Class DefinitionBase is made abstract.
